### PR TITLE
Wrap partitioned table factory and add Partition aggregation

### PR DESCRIPTION
**Pull request by jmao-denver**
Fixes: 2434

### CODE REVIEW

File: py/server/deephaven/table.py [hunk 1]
```diff
@@ -1343,6 +1345,76 @@ def __init__(self, j_partitioned_table):
...
+    def from_partitioned_table(cls,
+                               table: Table,
+                               key_cols: Union[str, List[str]] = None,
+                               unique_keys: bool = None,
+                               constituent_column: str = None,
+                               constituent_table_columns: List[Column] = None,
+                               constituent_changes_permitted: bool = None) -> PartitionedTable:
+        """Creates a PartitionedTable from the provided underlying partitioned Table.
+
+        Note: key_cols, unique_keys, constituent_table_columns, constituent_table_columns,
```
Comment (rcaudy) [inline py/server/deephaven/table.py:1358]:
```suggestion
        Note: key_cols, unique_keys, constituent_column, constituent_table_columns,
```

File: py/server/deephaven/table.py [hunk 2]
```diff
@@ -1343,6 +1345,76 @@ def __init__(self, j_partitioned_table):
...
+            constituent_table_columns (list[Column]): the column definitions of the constituent table
+            constituent_changes_permitted (bool): whether the values of the constituent column can change
+
+        Returns:
+            a PartitionedTable
+
+        Raise:
+            DHError
+        """
+        none_args = [key_cols, unique_keys, constituent_table_columns, constituent_table_columns,
```
Comment (rcaudy) [inline py/server/deephaven/table.py:1375]:
```suggestion
        none_args = [key_cols, unique_keys, constituent_column, constituent_table_columns,
```

File: py/server/deephaven/table.py [hunk 3]
```diff
@@ -1343,6 +1345,76 @@ def __init__(self, j_partitioned_table):
...
+                     constituent_changes_permitted]
+
+        try:
+            if all([arg is None for arg in none_args]):
+                return PartitionedTable(j_partitioned_table=_JPartitionedTableFactory.of(table.j_table))
+
+            if all([arg is not None for arg in none_args]):
+                table_def = _JTableDefinition.of([col.j_column_definition for col in constituent_table_columns])
+                j_partitioned_table = _JPartitionedTableFactory.of(table.j_table, j_array_list(to_sequence(key_cols)),
```
Comment (rcaudy) [inline py/server/deephaven/table.py:1384]: Formatting seems inconsistent on this line. Why put two args on this line, but not subsequent ones?

File: py/server/deephaven/table.py [hunk 4]
```diff
@@ -1343,6 +1345,76 @@ def __init__(self, j_partitioned_table):
...
+
+        Note: key_cols, unique_keys, constituent_table_columns, constituent_table_columns,
+        constituent_changes_permitted must either be all None or all have values
+
+        Args:
+            table (Table): the underlying partitioned table
+            key_cols (Union[str, List[str]]): the key column name(s) of 'table'
+            unique_keys (bool): whether the keys in 'table' are guaranteed to be unique
+            constituent_column (str): the constituent column name in 'table'
+            constituent_table_columns (list[Column]): the column definitions of the constituent table
```
Comment (rcaudy) [inline py/server/deephaven/table.py:1366]: I think both methods need to port more of the guidance from the Java version about requiring non-empty input in `table` if we need to infer the other args.

File: py/server/deephaven/table.py [hunk 5]
```diff
@@ -1343,6 +1345,76 @@ def __init__(self, j_partitioned_table):
...
+    @classmethod
+    def from_constituent_tables(cls,
+                                tables: List[Table],
+                                constituent_table_columns: List[Column] = None) -> PartitionedTable:
+        """Creates a PartitionedTable with a single column named '__CONSTITUENT__' containing the provided constituent
+        tables.
+
+        Args:
+            tables (List[Table]): the constituent tables
+            constituent_table_columns (List[Column]): the mutually compatible column definitions with all the
```
Comment (rcaudy) [inline py/server/deephaven/table.py:1405]: I think this documentation reads kind of weirdly.

**Review by rcaudy — DISMISSED** [review]

**Review by chipkent — DISMISSED** [review]

File: py/server/deephaven/table.py [hunk 6]
```diff
@@ -1343,6 +1345,95 @@ def __init__(self, j_partitioned_table):
...
+    def from_constituent_tables(cls,
+                                tables: List[Table],
+                                constituent_table_columns: List[Column] = None) -> PartitionedTable:
+        """Creates a PartitionedTable with a single column named '__CONSTITUENT__' containing the provided constituent
+        tables.
+
+        The result PartitionedTable has no key columns, both its unique_keys and constituent_changes_permitted
+        properties are set to False. When constituent_table_columns isn't provided, it will be set to the column
+        definitions of the first table in the provided constituent tables.
+
```
[the remark this answers is not in the document]
Comment (jmao-denver) [reply py/server/deephaven/table.py:1415]: an InCompatibleTableDefinition exception will be raised

**Review by rcaudy — APPROVED** [review]

**Pull request merged**